Web Inspector: Elements: add an editor for CSS `steps()`

dcrousso opened the pull request:
#### b56ed87ee07c4b8c343fa667da061b9cc96b2b73
<pre>
Web Inspector: Elements: add an editor for CSS `steps()`
<a href="https://bugs.webkit.org/show_bug.cgi?id=259567">https://bugs.webkit.org/show_bug.cgi?id=259567</a>

Reviewed by Patrick Angle.

We have visual previews/editors for `cubic-bezier()` and `spring()`, so we should also have something for `steps()`.

This will help developers visualize the behavior of their `steps()` function, as well as allow for better editing since there&apos;s an immediate visual feedback/preview.

* Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.js: Added.
(WI.StepsTimingFunctionEditor):
(WI.StepsTimingFunctionEditor.prototype.get element):
(WI.StepsTimingFunctionEditor.prototype.set stepsTimingFunction):
(WI.StepsTimingFunctionEditor.prototype.get stepsTimingFunction):
(WI.StepsTimingFunctionEditor.prototype._updateStepsTimingFunction):
(WI.StepsTimingFunctionEditor.prototype._updatePreview):
(WI.StepsTimingFunctionEditor.prototype._triggerPreviewAnimation):
(WI.StepsTimingFunctionEditor.prototype._resetPreviewAnimation):
(WI.StepsTimingFunctionEditor.prototype._handleTypeSelectElementChanged):
(WI.StepsTimingFunctionEditor.prototype._handleCountInputElementInput):
* Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.css: Added.
(.steps-timing-function-editor):
(.steps-timing-function-editor &gt; .preview):
(.steps-timing-function-editor &gt; .preview &gt; div):
(.steps-timing-function-editor &gt; .timing):
(.steps-timing-function-editor &gt; .preview.animate &gt; div, .steps-timing-function-editor &gt; .timing.animate):
(.steps-timing-function-editor &gt; svg):
(.steps-timing-function-editor &gt; svg path):
(.steps-timing-function-editor &gt; .parameter-editors):
(.steps-timing-function-editor &gt; .parameter-editors &gt; input):

* Source/WebInspectorUI/UserInterface/Models/TextMarker.js:
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorTextMarkers.js:
(createCodeMirrorStepsTimingFunctionTextMarkers): Added.
* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorStepsTimingFunctionEditingController.js: Added.
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.get initialValue):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.get cssClassName):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.popoverWillPresent):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.popoverDidPresent):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype._handleStepsTimingFunctionEditorStepsTimingFunctionChanged):
* Source/WebInspectorUI/UserInterface/Views/AnimationDetailsSidebarPanel.js:
(WI.AnimationDetailsSidebarPanel.prototype._refreshEffectSection):
* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor.prototype.createStepsTimingFunctionMarkers): Added.
(WI.TextEditor.prototype.editingControllerForMarker):
* Source/WebInspectorUI/UserInterface/Views/SourceCodeTextEditor.js:
(WI.SourceCodeTextEditor.prototype._updateEditableMarkers):
Add `WI.TextMarker.Type.StepsTimingFunction`.
Drive-by: Sort `WI.TextMarker.Type`.

* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js:
(WI.InlineSwatch):
(WI.InlineSwatch.prototype.didDismissPopover):
(WI.InlineSwatch.prototype._fallbackValue):
(WI.InlineSwatch.prototype._swatchElementClicked):
(WI.InlineSwatch.prototype._valueEditorValueDidChange):
* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.css:
(.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .variable)`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable)`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):hover): Renamed from `.inline-swatch:not(.read-only):is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable, .alignment):hover`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):active): Renamed from `.inline-swatch:not(.read-only):is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable, .alignment):active`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable) &gt; span): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable) &gt; span`.
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype._replaceSpecialTokens):
(WI.SpreadsheetStyleProperty.prototype._addTimingFunctionTokens):
(WI.SpreadsheetStyleProperty.prototype._addVariableTokens):
Add `WI.InlineSwatch.Type.StepsTimingFunction`.
Drive-by: Sort `WI.InlineSwatch.Type`.

* Source/WebInspectorUI/.eslintrc:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

Canonical link: <a href="https://commits.webkit.org/267484@main">https://commits.webkit.org/267484@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8579d93b95d2634a26ab710540e5b3a863f8b8a8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16728 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17052 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17502 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18510 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15671 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20284 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17191 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17967 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16927 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17298 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/14476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19298 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/14542 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/15160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15531 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/15326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19625 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15930 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15109 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4004 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19473 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15761 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebInspectorUI/UserInterface/Views/InlineSwatch.css:
```diff
@@ -50,19 +50,19 @@
 }
 
-.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .variable) {
+.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable) {
```
I feel like timing-functions could deserve their own shared class

dcrousso replied:
i personally like the explicitness of this as it means i can do a code search for `cubic-bezier-timing-function` and find all the various places in CSS that i might need to make changes (which i actually did multiple times for this PR as i wasn't immediately aware of all the places where i might need to support `steps()` stuff)

rcaliman-apple replied:
I tend to agree with Devin here. Abstraction is nice, but the ability to do global search-and-replace for specific functionality in a project the size and complexity of Web Inspector is even nicer :)

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.css:
```diff
@@ -0,0 +1,88 @@
+ */
+
+.steps-timing-function-editor {
+    width: 200px; /* Keep in sync with `editorWidth`. */
```
What if we replace the values for `width`, `margin`, `strokeWidth` and others that need to be kept in sync with CSS Custom Properties that `WI.StepsFunctionTimingEditor` sets on `this._element`? 

That way, the JavaScript implementation is responsible for setting these values for the stylesheet, and we don't have to remember to update them in both places.

```css
width: var(--steps-timing-function-editor-width);
```
In `WI.StepsFunctionTimingEditor`:
```js
this._element.style = `--steps-timing-function-editor-width: ${width}px`
```

Or vice-versa if you prefer, where `WI.StepsFunctionTimingEditor` reads the values of Custom Properties from the computed styles of `<div class="steps-timing-function-editor">` and uses them in JavaScript (though that's perhaps slightly more cumbersome with the need to defend JS from unexpected CSS values)

dcrousso replied:
ive sometimes found that a bit annoying to deal with because it now means that i have to look in another file to understand exactly what this file does, but generally speaking i do think the benefit of only having one source of truth is beneficial (not to mention that the variable name should clarify its intention, so even though the value isn't right here it can still be inferred)

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.js:
```diff
@@ -0,0 +1,209 @@
+            break;
+        }
+
+        let stepCount = this._stepsTimingFunction.count + stepCountAdjust;
+        let stepX = this._previewWidth / this._stepsTimingFunction.count; // Always use the defined number of steps to divide the duration.
+        let stepY = this._previewHeight / stepCount;
+
+        let path = [];
+        for (let i = stepStartAdjust; i <= stepCount; ++i) {
+            let x = stepX * (i - stepStartAdjust);
+            let y = this._previewHeight - (stepY * i);
+
+            if (goUpFirst) {
+                if (i)
+                    path.push("H", x);
+
+                if (i < stepCount)
+                    path.push("V", y - stepY);
+            } else {
+                path.push("V", y);
+
+                if (i < stepCount || stepCountAdjust < 0)
+                    path.push("H", x + stepX);
+            }
+        }
+        this._pathElement.setAttribute("d", `M 0 ${this._previewHeight} ${path.join(" ")} L ${this._previewWidth} 0`);
```
This is a nit, but when there are more than ~200 steps, the SVG illustration practically turns into a straight line. It doesn't convey any visual information to render it at as a series of steps. 

<img width="329" alt="Screenshot 2023-07-31 at 15 34 38" src="https://github.com/WebKit/WebKit/assets/105809798/a5fa55ed-298f-448c-94e9-85793246ab05">


Performance-wise it only becomes an issue for extreme cases, like 1,000,000 steps, where Web Inspector's page gets allocated a lot of memory (1Gb+ on my test) to render the needlessly complex SVG path.


<img width="328" alt="Screenshot 2023-07-31 at 15 33 14" src="https://github.com/WebKit/WebKit/assets/105809798/b8d5e3dc-0e45-4f57-b724-371e8001ab50">

Of course, this shouldn't occur in practice (famous last words 🙂), but if it's straightforward, perhaps we can alter the illustration to a straight line beyond a reasonable number steps and mitigate this. Again, this is a nit.

dcrousso replied:
lol interesting find.  yeah we probably could optimize this to not do that when there's enough steps, but IMO that's probably not something super common so i think we can wait to do that until it's necessary (or at least in a followup)

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.js:
```diff
@@ -0,0 +1,209 @@
+
+        this._stepsTimingFunction = stepsTimingFunction;
+
+        this._typeSelectElement.value = this._stepsTimingFunction.type;
```
When the timing function value has just a step count and is missing a step position, like `steps(5)`, the editor defaults to `WI.StepsTimingFunction.Type.JumpEnd` and inserts this into the output when using the tool to change the step count: `steps(5, jump-end)`.

I noticed now that the default value we set in [`WI.StepsTimingFunction.fromString()`](https://github.com/WebKit/WebKit/blob/3ed63da0ea8dcf64cb82bd1a61e79c38a2aeaae0/Source/WebInspectorUI/UserInterface/Models/StepsTimingFunction.js#L56) for the step position is `WI.StepsTimingFunction.Type.JumpEnd` 

```
let type = matches[2] || WI.StepsTimingFunction.Type.JumpEnd;
```

The spec says that `end` [is assumed if omitted](https://www.w3.org/TR/css-easing-1/#funcdef-step-easing-function-steps) (which is equivalent to `jump-end`):

>If the second parameter is omitted, the value [end](https://www.w3.org/TR/css-easing-1/#valdef-steps-end) is assumed.

Interestingly, the [serialization rules](https://www.w3.org/TR/css-easing-1/#serialization) say that values like `steps(5, jump-end)` and `steps(5, end)` serialize to just `steps(5)`. WebKit respects this when returning computed styles.

I don't have a strong opinion either way, given the end result is identical in this scenario, but should the editor  modify the authored value to insert a default step position if there wasn't one to begin with? I.e.: should the editor strive to keep `steps(5)` unless the user explicitly uses the select menu to pick a value for step position?

dcrousso replied:
generally speaking i think Web Inspector tries to follow the input format as much as possible, so yeah i think in this case we could make it output the most efficient form whenever possible.  i can do that as a followup :)

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.js:
```diff
@@ -0,0 +1,209 @@
+        this._element.classList.add("steps-timing-function-editor");
+        this._element.dir = "ltr";
+
+        const strokeWidth = 2; // Keep in sync with `.steps-timing-function-editor > svg path`.
```
See other comment from `StepsTimingFunctionEditor.css` about perhaps sharing these values between JS and CSS with CSS Custom Properties.

rcaliman-apple reviewed (approved):
This looks good to me. 👍

Thank you for adding a specialized editor and preview for the `steps()` timing function!
I tried it out on a few examples and edge cases, and it performs great!

Left a suggestion about sharing values between CSS and JS with CSS Custom Properties and a few questions. Nothing blocking. Great work!

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/InlineSwatch.css:
```diff
@@ -50,19 +50,19 @@
 }
 
-.inline-swatch:is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable) {
+.inline-swatch:is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .steps-timing-function, .variable) {
```
Any chance you want to alphabetize these while you are here?

patrickangle reviewed (approved):
The UI doesn’t quite feel right to me since it lacks any sense of where the graph area begins or ends… As a follow up we might want to consider bringing some clarity to these various timing popovers, perhaps even as simple as a white/primary content background color in the actual graph area?

dcrousso committed:
Web Inspector: Elements: add an editor for CSS `steps()`
https://bugs.webkit.org/show_bug.cgi?id=259567

Reviewed by Patrick Angle.

We have visual previews/editors for `cubic-bezier()` and `spring()`, so we should also have something for `steps()`.

This will help developers visualize the behavior of their `steps()` function, as well as allow for better editing since there's an immediate visual feedback/preview.

* Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.js: Added.
(WI.StepsTimingFunctionEditor):
(WI.StepsTimingFunctionEditor.prototype.get element):
(WI.StepsTimingFunctionEditor.prototype.set stepsTimingFunction):
(WI.StepsTimingFunctionEditor.prototype.get stepsTimingFunction):
(WI.StepsTimingFunctionEditor.prototype._updateStepsTimingFunction):
(WI.StepsTimingFunctionEditor.prototype._updatePreview):
(WI.StepsTimingFunctionEditor.prototype._triggerPreviewAnimation):
(WI.StepsTimingFunctionEditor.prototype._resetPreviewAnimation):
(WI.StepsTimingFunctionEditor.prototype._handleTypeSelectElementChanged):
(WI.StepsTimingFunctionEditor.prototype._handleCountInputElementInput):
* Source/WebInspectorUI/UserInterface/Views/StepsTimingFunctionEditor.css: Added.
(.steps-timing-function-editor):
(.steps-timing-function-editor > .preview):
(.steps-timing-function-editor > .preview > div):
(.steps-timing-function-editor > .timing):
(.steps-timing-function-editor > .preview.animate > div, .steps-timing-function-editor > .timing.animate):
(.steps-timing-function-editor > svg):
(.steps-timing-function-editor > svg path):
(.steps-timing-function-editor > .parameter-editors):
(.steps-timing-function-editor > .parameter-editors > input):

* Source/WebInspectorUI/UserInterface/Models/TextMarker.js:
* Source/WebInspectorUI/UserInterface/Views/CodeMirrorTextMarkers.js:
(createCodeMirrorStepsTimingFunctionTextMarkers): Added.
* Source/WebInspectorUI/UserInterface/Controllers/CodeMirrorStepsTimingFunctionEditingController.js: Added.
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.get initialValue):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.get cssClassName):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.popoverWillPresent):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype.popoverDidPresent):
(WI.CodeMirrorStepsTimingFunctionEditingController.prototype._handleStepsTimingFunctionEditorStepsTimingFunctionChanged):
* Source/WebInspectorUI/UserInterface/Views/AnimationDetailsSidebarPanel.js:
(WI.AnimationDetailsSidebarPanel.prototype._refreshEffectSection):
* Source/WebInspectorUI/UserInterface/Views/TextEditor.js:
(WI.TextEditor.prototype.createStepsTimingFunctionMarkers): Added.
(WI.TextEditor.prototype.editingControllerForMarker):
* Source/WebInspectorUI/UserInterface/Views/SourceCodeTextEditor.js:
(WI.SourceCodeTextEditor.prototype._updateEditableMarkers):
Add `WI.TextMarker.Type.StepsTimingFunction`.
Drive-by: Sort `WI.TextMarker.Type`.

* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js:
(WI.InlineSwatch):
(WI.InlineSwatch.prototype.didDismissPopover):
(WI.InlineSwatch.prototype._fallbackValue):
(WI.InlineSwatch.prototype._swatchElementClicked):
(WI.InlineSwatch.prototype._valueEditorValueDidChange):
* Source/WebInspectorUI/UserInterface/Views/InlineSwatch.css:
(.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .spring-timing-function, .variable)`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable)): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable)`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):hover): Renamed from `.inline-swatch:not(.read-only):is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable, .alignment):hover`.
(.inline-swatch:not(.read-only):is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable, .alignment):active): Renamed from `.inline-swatch:not(.read-only):is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable, .alignment):active`.
(.inline-swatch:is(.box-shadow, .cubic-bezier-timing-function, .spring-timing-function, .steps-timing-function, .variable) > span): Renamed from `.inline-swatch:is(.cubic-bezier-timing-function, .box-shadow, .spring-timing-function, .variable) > span`.
* Source/WebInspectorUI/UserInterface/Views/SpreadsheetStyleProperty.js:
(WI.SpreadsheetStyleProperty.prototype._replaceSpecialTokens):
(WI.SpreadsheetStyleProperty.prototype._addTimingFunctionTokens):
(WI.SpreadsheetStyleProperty.prototype._addVariableTokens):
Add `WI.InlineSwatch.Type.StepsTimingFunction`.
Drive-by: Sort `WI.InlineSwatch.Type`.

* Source/WebInspectorUI/.eslintrc:
* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

Canonical link: https://commits.webkit.org/267484@main
